File: src/proxy/response_builder.rs
//! Response building utilities for proxy responses.
//!
//! Handles the construction of `ProxyResponse` from raw HTTP response data,
//! including decompression, binary detection, and size calculations.

use super::types::*;
use crate::infra::decompress_body;
use crate::shared::{status_text, CapturedCertInfo, DetailedTiming};
use base64::Engine;
use hyper::Version;
use std::collections::HashMap;
use std::net::IpAddr;

/// Determines if response body is likely binary based on content-type.
///
/// # Arguments
///
/// * `content_type` - The Content-Type header value
///
/// # Returns
///
/// `true` if the content is likely binary, `false` otherwise.
pub fn is_binary_content(content_type: Option<&str>) -> bool {
    let ct = match content_type {
        Some(ct) => ct.to_lowercase(),
        None => return false,
    };

    let text_types = [
        "text/",
        "application/json",
        "application/xml",
        "application/javascript",
        "application/x-javascript",
        "application/ecmascript",
        "application/x-www-form-urlencoded",
        "+json",
        "+xml",
    ];

    !text_types.iter().any(|t| ct.contains(t))
}

/// Converts an HTTP version to its string representation.
pub fn version_to_string(version: Version) -> String {
    match version {
        Version::HTTP_09 => "HTTP/0.9",
        Version::HTTP_10 => "HTTP/1.0",
        Version::HTTP_11 => "HTTP/1.1",
        Version::HTTP_2 => "HTTP/2",
        Version::HTTP_3 => "HTTP/3",
        _ => "HTTP/1.1",
    }
    .to_string()
}

/// Parameters for building a proxy response.
pub struct ResponseBuildParams {
    pub status: u16,
    pub headers: HashMap<String, String>,
    pub body_bytes: Vec<u8>,
    pub timing: DetailedTiming,
    pub final_url: String,
    pub redirect_chain: Vec<RedirectHop>,
    pub tls_info: Option<CapturedCertInfo>,
    pub http_version: Version,
    pub server_ip: Option<IpAddr>,
    pub request_headers: HashMap<String, String>,
    pub request_body_size: Option<usize>,
}

/// Builds a `ProxyResponse` from raw response data.
///
/// Handles:
/// - Body decompression (gzip, deflate, brotli)
/// - Binary content detection and base64 encoding
/// - Size breakdown calculation
/// - TLS info conversion
///
/// # Arguments
///
/// * `params` - The response build parameters
///
/// # Returns
///
/// A `ProxyResponse` ready to be serialized and sent to the client.
pub fn build_response(params: ResponseBuildParams) -> ProxyResponse {
    let ResponseBuildParams {
        status,
        headers,
        body_bytes,
        timing,
        final_url,
        redirect_chain,
        tls_info,
        http_version,
        server_ip,
        request_headers,
        request_body_size,
    } = params;

    let content_type = headers.get("content-type").map(|s| s.as_str());
    let content_encoding = headers.get("content-encoding").map(|s| s.as_str());
    let is_binary = is_binary_content(content_type);

    // Decompress if needed
    let compressed_size = body_bytes.len();
    let decompressed: Vec<u8> = match decompress_body(&body_bytes, content_encoding) {
        Ok(d) => d,
        Err(e) => {
            return ProxyResponse::error(e, "DECOMPRESSION_ERROR".to_string());
        }
    };
    let body_size = decompressed.len();

    // Convert body
    let (body, body_base64) = if is_binary {
        let b64 = base64::engine::general_purpose::STANDARD.encode(&decompressed);
        (String::new(), Some(b64))
    } else {
        (String::from_utf8_lossy(&decompressed).to_string(), None)
    };

    // Calculate sizes
    let status_line = format!(
        "{} {} {}",
        version_to_string(http_version),
        status,
        status_text(status)
    );
    let header_size: usize = headers
        .iter()
        .map(|(k, v)| k.len() + 2 + v.len() + 2)
        .sum::<usize>()
        + status_line.len()
        + 2;

    let compression_ratio = if content_encoding.is_some() && body_size > 0 {
        Some(compressed_size as f64 / body_size as f64)
    } else {
        None
    };

    let size_breakdown = SizeBreakdown {
        headers: header_size,
        body: body_size,
        total: header_size + body_size,
        compressed: if content_encoding.is_some() {
            Some(compressed_size)
        } else {
            None
        },
        uncompressed: if content_encoding.is_some() {
            Some(body_size)
        } else {
            None
        },
        encoding: content_encoding.map(|s| s.to_string()),
        compression_ratio,
    };

    // Build TLS info
    let tls = tls_info.map(|info| TlsInfo {
        protocol: Some(info.protocol),
        cipher: Some(info.cipher),
        issuer: info.issuer,
        subject: info.subject,
        valid_from: info.valid_from,
        valid_to: info.valid_to,
        valid: Some(true),
    });

    let server_software = headers.get("server").cloned();
    let connection = headers.get("connection").cloned();

    let data = ResponseData {
        status,
        status_text: status_text(status),
        headers,
        request_headers: Some(request_headers),
        body,
        body_base64,
        is_binary,
        size: body_size,
        timing: timing.to_timing_info(),
        url: final_url,
        redirected: !redirect_chain.is_empty(),
        redirect_chain: if redirect_chain.is_empty() {
            None
        } else {
            Some(redirect_chain)
        },
        tls,
        size_breakdown: Some(size_breakdown),
        server_ip: server_ip.map(|ip| ip.to_string()),
        protocol: Some(version_to_string(http_version)),
        from_cache: Some(false),
        resource_type: Some("fetch".to_string()),
        request_body_size,
        connection,
        server_software,
    };

    ProxyResponse::success(data)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_is_binary_content() {
        assert!(!is_binary_content(Some("text/html")));
        assert!(!is_binary_content(Some("application/json")));
        assert!(!is_binary_content(Some("application/xml")));
        assert!(!is_binary_content(Some("text/plain; charset=utf-8")));
        assert!(is_binary_content(Some("image/png")));
        assert!(is_binary_content(Some("application/octet-stream")));
        assert!(!is_binary_content(None));
    }

    #[test]
    fn test_version_to_string() {
        assert_eq!(version_to_string(Version::HTTP_11), "HTTP/1.1");
        assert_eq!(version_to_string(Version::HTTP_2), "HTTP/2");
        assert_eq!(version_to_string(Version::HTTP_10), "HTTP/1.0");
    }
}

File: internal/proxy/response_builder.go
package proxy

import (
	"encoding/base64"
	"fmt"
	"strings"
	"time"

	"zone.digit.tommie/internal/infra"
	"zone.digit.tommie/pkg/statustext"
)

// responseBuildParams contains parameters for building a proxy response.
type responseBuildParams struct {
	status          uint16
	headers         map[string]string
	bodyBytes       []byte
	timing          *DetailedTiming
	finalURL        string
	redirectChain   []RedirectHop
	tlsInfo         *infra.CertInfo
	httpVersion     string
	serverIP        string
	requestHeaders  map[string]string
	requestBodySize *int
	hostname        string
	port            string
	resolvedIPs     []string
}

// isBinaryContent determines if response body is likely binary based on content-type.
func isBinaryContent(contentType string) bool {
	if contentType == "" {
		return false
	}

	ct := strings.ToLower(contentType)

	textTypes := []string{
		"text/",
		"application/json",
		"application/xml",
		"application/javascript",
		"application/x-javascript",
		"application/ecmascript",
		"application/x-www-form-urlencoded",
		"+json",
		"+xml",
	}

	for _, t := range textTypes {
		if strings.Contains(ct, t) {
			return false
		}
	}

	return true
}

// buildResponse builds a ProxyResponse from raw response data.
func buildResponse(params responseBuildParams) ProxyResponse {
	contentType := params.headers["content-type"]
	contentEncoding := params.headers["content-encoding"]
	isBinary := isBinaryContent(contentType)

	// Decompress if needed
	compressedSize := len(params.bodyBytes)
	decompressResult, err := infra.Decompress(params.bodyBytes, contentEncoding)
	if err != nil {
		return NewErrorResponse(fmt.Sprintf("Decompression failed: %v", err), "DECOMPRESSION_ERROR")
	}
	decompressed := decompressResult.Data
	bodySize := len(decompressed)

	// Convert body
	var body string
	var bodyBase64 *string
	if isBinary {
		b64 := base64.StdEncoding.EncodeToString(decompressed)
		bodyBase64 = &b64
	} else {
		body = string(decompressed)
	}

	// Calculate sizes
	statusLine := fmt.Sprintf("%s %d %s", params.httpVersion, params.status, statustext.Get(int(params.status)))
	headerSize := len(statusLine) + 2
	for k, v := range params.headers {
		headerSize += len(k) + 2 + len(v) + 2
	}

	var compressionRatio *float64
	var compressed *int
	var uncompressed *int
	var encoding *string

	if contentEncoding != "" && bodySize > 0 {
		ratio := float64(compressedSize) / float64(bodySize)
		compressionRatio = &ratio
		compressed = &compressedSize
		uncompressed = &bodySize
		encoding = &contentEncoding
	}

	sizeBreakdown := &SizeBreakdown{
		Headers:          headerSize,
		Body:             bodySize,
		Total:            headerSize + bodySize,
		Compressed:       compressed,
		Uncompressed:     uncompressed,
		Encoding:         encoding,
		CompressionRatio: compressionRatio,
	}

	// Build TLS info
	var tlsInfoData *TLSInfo
	if params.tlsInfo != nil {
		valid := infra.IsCertValid(params.tlsInfo.ValidFrom, params.tlsInfo.ValidTo)
		tlsInfoData = &TLSInfo{
			Protocol:  strPtr(params.tlsInfo.Protocol),
			Cipher:    strPtr(params.tlsInfo.Cipher),
			Issuer:    strPtr(params.tlsInfo.Issuer),
			Subject:   strPtr(params.tlsInfo.Subject),
			ValidFrom: &params.tlsInfo.ValidFrom,
			ValidTo:   &params.tlsInfo.ValidTo,
			Valid:     &valid,
			SANs:      params.tlsInfo.SANs,
		}
	}

	serverSoftware := params.headers["server"]
	connection := params.headers["connection"]

	var serverSoftwarePtr, connectionPtr *string
	if serverSoftware != "" {
		serverSoftwarePtr = &serverSoftware
	}
	if connection != "" {
		connectionPtr = &connection
	}

	var serverIPPtr *string
	if params.serverIP != "" {
		serverIPPtr = &params.serverIP
	}

	var hostnamePtr *string
	if params.hostname != "" {
		hostnamePtr = &params.hostname
	}

	var portPtr *string
	if params.port != "" {
		portPtr = &params.port
	}

	fromCache := false
	resourceType := "fetch"

	var redirectChainPtr []RedirectHop
	if len(params.redirectChain) > 0 {
		redirectChainPtr = params.redirectChain
	}

	data := ResponseData{
		Status:          params.status,
		StatusText:      statustext.Get(int(params.status)),
		Headers:         params.headers,
		RequestHeaders:  params.requestHeaders,
		Body:            body,
		BodyBase64:      bodyBase64,
		IsBinary:        isBinary,
		Size:            bodySize,
		Timing:          params.timing.ToTimingInfo(),
		URL:             params.finalURL,
		Redirected:      len(params.redirectChain) > 0,
		RedirectChain:   redirectChainPtr,
		TLS:             tlsInfoData,
		SizeBreakdown:   sizeBreakdown,
		ServerIP:        serverIPPtr,
		Protocol:        &params.httpVersion,
		FromCache:       &fromCache,
		ResourceType:    &resourceType,
		RequestBodySize: params.requestBodySize,
		Connection:      connectionPtr,
		ServerSoftware:  serverSoftwarePtr,
		Hostname:        hostnamePtr,
		Port:            portPtr,
		ResolvedIPs:     params.resolvedIPs,
	}

	return NewSuccessResponse(data)
}

// uint64Ptr creates a pointer to a uint64.
func uint64Ptr(v uint64) *uint64 {
	return &v
}

// nowUnix returns the current Unix timestamp.
func nowUnix() uint64 {
	return uint64(time.Now().Unix())
}
